src/rust/engine/src/nodes.rs

```diff
@@ -1053,22 +1053,25 @@ impl Node for NodeKey {
 
   fn run(self, context: Context) -> NodeFuture<NodeResult> {
     let mut workunit_state = workunit_store::expect_workunit_state();
-    let maybe_started_workunit_id: Option<String> = if context.session.should_handle_workunits() {
-      self.user_facing_name().map(|node_name| {
-        let span_id = new_span_id();
-        let desc = self.display_info().and_then(|di| di.desc.as_ref().cloned());
 
-        // We're starting a new workunit: record our parent, and set the current parent to our span.
-        let parent_id = std::mem::replace(&mut workunit_state.parent_id, Some(span_id.clone()));
-        let metadata = WorkunitMetadata { desc };
+    let started_workunit_id = {
+      let display = context.session.should_handle_workunits() && self.user_facing_name().is_some();
```

Comment:
It seems like `context.session.should_handle_workunits()` is now only recorded as a boolean... that might be fine, but we should also maybe just consider removing it?

The two potentially expensive parts are 1) recording the workunits (here), and 2) actually adding them to the runtracker on the Python side. It's possible that moving that flag to only skip the adding of workunits to the runtracker would eliminate any differences.

Reply:
Agreed, we should rethink this, but I'd also like to save that for another PR.

Reply:
Well, you're changing it here: we used to not record workunits if the flag wasn't set: now we do. Should at least confirm that this change doesn't make things slower before landing.

Reply:
Running a few pants `test` goals on both this PR branch and master doesn't seem to show much of a difference in timing:

```
This PR

time ./v2 test src/python/pants/engine::
________________________________________________________
Executed in   83.27 secs   fish           external 
   usr time  240.45 secs  1404.00 micros  240.45 secs 
   sys time   27.92 secs  418.00 micros   27.91 secs 

time ./v2 test tests/python/pants_test/util::
________________________________________________________
Executed in    6.90 secs   fish           external 
   usr time    8.16 secs  817.00 micros    8.16 secs 
   sys time    1.55 secs  244.00 micros    1.55 secs 




Master:

time ./v2 test src/python/pants/engine::
________________________________________________________
Executed in   83.93 secs   fish           external 
   usr time  250.31 secs  1144.00 micros  250.31 secs 
   sys time   28.81 secs  342.00 micros   28.81 secs 


time ./v2 test tests/python/pants_test/util::
________________________________________________________
Executed in    6.64 secs   fish           external 
   usr time    7.73 secs    0.00 micros    7.73 secs 
   sys time    1.67 secs  1448.00 micros    1.67 secs 
```